### PR TITLE
reconstruct JSON code

### PR DESCRIPTION
**Pull request by hicqu**

### CODE REVIEW

**Comment by hicqu** [comment]
reimplement JSON. PTAL, thanks!

File: util/types/json/compare.go [hunk 1]
```diff
@@ -15,11 +15,11 @@ package json
 
 import (
 	"bytes"
-	"reflect"
 	"strings"
+	"unsafe"
 )
 
-// floatEpsilon is for compare two float value allowing precision loss.
+// floatEpsilon is the acceptable error precision when comparing two floating numbers.
```
Comment (zhexuany) [inline util/types/json/compare.go]: the acceptable error quantity when comparing two float numbers.

File: util/types/json/compare.go [hunk 2]
```diff
@@ -51,31 +51,15 @@ var jsonTypePrecedences = map[string]int{
...
+		return float64(*bit64)
+	case typeCodeFloat64:
+		return *(*float64)(unsafe.Pointer(bit64))
+	default:
+		panic(internalErrorUnknownTypeCode)
 	}
-	return string(j.(jsonString))
 }
 
 // CompareJSON compares two json object.
```
Comment (zhexuany) [inline util/types/json/compare.go]: remove this line since Json is gone.

File: util/types/json/serdes.go [hunk 3]
```diff
@@ -371,32 +342,14 @@ var jsonTypeCodeLength = map[byte]int{
...
-		return new(jsonInt64)
-	case typeCodeDouble:
-		return new(jsonDouble)
-	case typeCodeString:
-		return new(jsonString)
-	}
-	panic("unknown type code")
-}
-
 // Two map are equal if they have same keys and same values.
```
Comment (zhexuany) [inline util/types/json/serdes.go]: please add method name at the beginning of comment.

File: util/types/json/json.go [hunk 4]
```diff
@@ -15,26 +15,130 @@ package json
...
+// AccessMethod is for MyQL JSON functions.
+type AccessMethod interface {
+	// Type is for function JSON_TYPE.
 	Type() string
 }
 
+// JSON is for MySQL JSON type.
+type JSON struct {
+	typeCode byte
+	bit64    int64
```
Comment (shenli) [inline util/types/json/json.go]: i64 is a better name.

File: util/types/json/compare.go [hunk 5]
```diff
@@ -51,31 +51,15 @@ var jsonTypePrecedences = map[string]int{
...
-		}
-		return float64(j.(jsonLiteral))
-	}
-	panic("can not convert to float64")
-}
-
-func jsonAsString(j JSON) string {
-	if reflect.TypeOf(j).Kind() == reflect.Ptr {
-		return string(*j.(*jsonString))
+func bit64AsFloat64(bit64 *int64, typeCode byte) float64 {
```
Comment (shenli) [inline util/types/json/compare.go]: Why use pointer as the first argument?

File: util/types/json/serdes.go [hunk 6]
```diff
@@ -305,11 +280,7 @@ func (m *jsonObject) decode(data []byte) (err error) {
 	return
 }
 
-func (a jsonArray) getTypeCode() byte {
-	return typeCodeArray
-}
-
-func (a jsonArray) encode(buffer *bytes.Buffer) {
+func encodeJSONArray(a []JSON, buffer *bytes.Buffer) {
```
Comment (zhexuany) [inline util/types/json/serdes.go]: I do not think this is right place to do encoding. We should put encoding code to `util/codec` and add new type in `mysql` package. If you plan to do this in another PR, just ignore this comment.

File: util/types/json/json.go [hunk 7]
```diff
@@ -15,26 +15,130 @@ package json
...
+	jsonLiteralNil   byte = 0x00
+	jsonLiteralTrue  byte = 0x01
+	jsonLiteralFalse byte = 0x02
+)
+
+const internalErrorUnknownTypeCode = "unknown type code"
+
+var _ AccessMethod = JSON{}
+
+// AccessMethod is for MyQL JSON functions.
```
Comment (shenli) [inline util/types/json/json.go]: MyQL -> MySQL

File: util/types/json/serdes.go [hunk 8]
```diff
@@ -258,7 +233,7 @@ func (m jsonObject) encode(buffer *bytes.Buffer) {
 	return
 }
```
Comment (zhexuany) [inline util/types/json/serdes.go]: same reason as I said in encoding part.

File: util/types/json/json.go (hunk 7 above)
Comment (shenli) [inline util/types/json/json.go]: Why should we need this interface?

File: util/types/json/serdes.go (hunk 6 above)
Comment (hicqu) [inline util/types/json/serdes.go]: package util/codec cannot access JSON internal fields, so I put serialize/deserialize here.
and, codec in util/codec will use these interfaces.

File: util/types/json/json.go (hunk 7 above)
Comment (hicqu) [inline util/types/json/json.go]: it's not required. I use it just as a method list, so we can know which access methods JSON has more easily.  Is that a bad style?

File: util/types/json/normalize.go [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+// limitations under the License.
+
+package json
+
+import (
+	"encoding/json"
+	"unsafe"
+)
+
+func normalize(in interface{}) (j JSON) {
```
Comment (shenli) [inline util/types/json/normalize.go]: Add comments for this function.

File: util/types/json/serdes.go [hunk 10]
```diff
@@ -371,32 +342,14 @@ var jsonTypeCodeLength = map[byte]int{
 	typeCodeArray:   -1,
 	typeCodeLiteral: 1,
 	typeCodeInt64:   8,
-	typeCodeDouble:  8,
+	typeCodeFloat64: 8,
```
Comment (shenli) [inline util/types/json/serdes.go]: Why change this?

Comment (hicqu) [inline util/types/json/serdes.go]: it's more symmetric with `Int64`

File: util/types/json/compare.go [hunk 11]
```diff
@@ -58,7 +59,8 @@ func i64AsFloat64(i64 int64, typeCode byte) float64 {
 	case typeCodeFloat64:
 		return *(*float64)(unsafe.Pointer(&i64))
 	default:
-		panic(internalErrorUnknownTypeCode)
+		msg := fmt.Sprintf(internalErrorUnknownTypeCode, typeCode)
```
Comment (shenli) [inline util/types/json/compare.go]: The name looks weird. How about `unknownTypeCodeErrorMsg`.

File: util/types/json/json.go (hunk 7 above)
Comment (shenli) [inline util/types/json/json.go]: This is nor necessary.

File: util/types/json/compare.go [hunk 12]
```diff
@@ -51,31 +52,16 @@ var jsonTypePrecedences = map[string]int{
...
+	case typeCodeFloat64:
+		return *(*float64)(unsafe.Pointer(&i64))
+	default:
+		msg := fmt.Sprintf(internalErrorUnknownTypeCode, typeCode)
+		panic(msg)
 	}
-	return string(j.(jsonString))
 }
 
 // CompareJSON compares two json object.
```
Comment (tiancaiamao) [inline util/types/json/compare.go]: Add comment for the meaning of returned cmp.

File: util/types/json/compare.go [hunk 13]
```diff
@@ -88,32 +74,33 @@ func CompareJSON(j1 JSON, j2 JSON) (cmp int, err error) {
...
+		switch j1.typeCode {
+		case typeCodeLiteral:
+			left := j1.i64
+			right := j2.i64
 			// false is less than true.
-			cmp = right - left
-		case jsonInt64, jsonDouble:
-			left := jsonAsFloat64(j1)
-			right := jsonAsFloat64(j2)
+			cmp = int(right - left)
```
Comment (tiancaiamao) [inline util/types/json/compare.go]: cmp > 0 when j1 < j2 ?

File: util/types/json/compare.go [hunk 14]
```diff
@@ -88,32 +74,33 @@ func CompareJSON(j1 JSON, j2 JSON) (cmp int, err error) {
...
-			for i := 0; i < len(x) && i < len(y); i++ {
-				elem1 := x[i]
-				elem2 := y[i]
+		case typeCodeArray:
+			left := j1.array
+			right := j2.array
+			for i := 0; i < len(left) && i < len(right); i++ {
+				elem1 := left[i]
+				elem2 := right[i]
 				cmp, _ = CompareJSON(elem1, elem2)
```
Comment (tiancaiamao) [inline util/types/json/compare.go]: why ignore error?

File: util/types/json/compare.go (hunk 13 above)
Comment (hicqu) [inline util/types/json/compare.go]: that's only for literal. for json literal, true is encoded as 0x01, but false as 0x02. we need true > false, so we use `right - left`.

File: util/types/json/json.go [hunk 15]
```diff
@@ -15,26 +15,126 @@ package json
...
+type JSON struct {
+	typeCode byte
+	i64      int64
+	str      string
+	object   map[string]JSON
+	array    []JSON
 }
 
 // CreateJSON creates a JSON from in. Return nil if in is not a valid JSON type.
 func CreateJSON(in interface{}) JSON {
```
Comment (tiancaiamao) [inline util/types/json/json.go]: if JSON is a big struct, it's better return `*JSON`, and define method on `*JSON`.

Comment (hicqu) [inline util/types/json/json.go]: JSON won't bigger than Datum, and we have many functions return Datum directly.
so I prefer return it on stack. What do you think?

File: util/types/json/json.go [hunk 16]
```diff
@@ -15,26 +15,126 @@ package json
...
 
+// ParseFromString parses a json from string.
+func ParseFromString(s string) (j JSON, err error) {
+	if len(s) == 0 {
+		err = ErrInvalidJSONText.GenByArgs("The document is empty")
+		return
+	}
+	var decoder = json.NewDecoder(bytes.NewReader(hack.Slice(s)))
+	decoder.UseNumber()
+	var in interface{}
```
Comment (tiancaiamao) [inline util/types/json/json.go]: // TODO: Implement the decoder directly would improve performance.

Comment (tiancaiamao) [inline util/types/json/json.go]: Maybe you can implement `json.Unmarshaler` on JSON

File: util/types/json/json.go [hunk 17]
```diff
@@ -15,26 +15,126 @@ package json
...
 	return normalize(in)
 }
 
+// ParseFromString parses a json from string.
+func ParseFromString(s string) (j JSON, err error) {
+	if len(s) == 0 {
+		err = ErrInvalidJSONText.GenByArgs("The document is empty")
+		return
+	}
+	var decoder = json.NewDecoder(bytes.NewReader(hack.Slice(s)))
```
Comment (tiancaiamao) [inline util/types/json/json.go]: use `bytes.NewBufferString(s)` replace `bytes.NewReader` ?

File: util/types/json/json.go [hunk 18]
```diff
@@ -15,26 +15,126 @@ package json
...
+func (j JSON) MarshalJSON() ([]byte, error) {
+	switch j.typeCode {
+	case typeCodeObject:
+		return json.Marshal(j.object)
+	case typeCodeArray:
+		return json.Marshal(j.array)
+	case typeCodeLiteral:
+		switch byte(j.i64) {
+		case jsonLiteralNil:
+			return hack.Slice("null"), nil
```
Comment (tiancaiamao) [inline util/types/json/json.go]: `[]byte("null")`
hack.Slick is black magic, don't use it too much!

File: util/types/json/json.go [hunk 19]
```diff
@@ -15,26 +15,133 @@ package json
...
+}
+
+// String implements fmt.Stringer interface.
+func (j JSON) String() string {
+	bytes, _ := json.Marshal(j)
+	return strings.TrimSpace(hack.String(bytes))
+}
+
+// Type returns type of JSON as string.
+func (j JSON) Type() string {
```
Comment (tiancaiamao) [inline util/types/json/json.go]:
```
type JSONType byte
const (
      typeCodeObject  JSONType = 0x01
      ...
)
func (j JSON) Type() JSONType
```

Comment (tiancaiamao) [inline util/types/json/json.go]: Compare string is bad for performance, compare JSONType is much faster.

File: util/types/json/json_test.go [hunk 20]
```diff
@@ -112,6 +112,7 @@ func (s *testJSONSuite) TestJSONType(c *C) {
 
 func (s *testJSONSuite) TestCompareJSON(c *C) {
 	var cmp int
+	var err error
 
 	// compare two JSON boolean.
 	jBool1, _ := ParseFromString(`true`)
```
Comment (tiancaiamao) [inline util/types/json/json_test.go]: You can make a table, rewrite the code to test case driven.

File: util/types/json/normalize.go [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+		}
+	case []interface{}:
+		j.typeCode = typeCodeArray
+		j.array = make([]JSON, 0, len(t))
+		for _, elem := range t {
+			j.array = append(j.array, normalize(elem))
+		}
+	default:
+		msg := fmt.Sprintf(unknownTypeErrorMsg, reflect.TypeOf(in))
+		panic(msg)
```
Comment (tiancaiamao) [inline util/types/json/normalize.go]: I prefer return error to panic. The input of this function is unknown.

File: util/types/json/serdes.go [hunk 22]
```diff
@@ -371,32 +345,12 @@ var jsonTypeCodeLength = map[byte]int{
...
-}
-
-// Two map are equal if they have same keys and same values.
-// So we sort the keys before serialize in order to keep
-// their binary representations are same.
-func (m jsonObject) getSortedKeys() []string {
+// getSortedKeys returns sorted keys of a map.
+func getSortedKeys(m map[string]JSON) []string {
 	keys := make([]string, 0, len(m))
 	for k := range m {
```
Comment (tiancaiamao) [inline util/types/json/serdes.go]: The order of  this operation is undetermined, call `getSortedKeys` multiple times you may get different result!

File: util/types/json/json.go (hunk 19 above)
Comment (hicqu) [inline util/types/json/json.go]: I prefer use byte, because I need `buffer.WriteByte(typeCodeObject)`. if typedef them to other names, I need coerce them to byte.

File: util/types/json/serdes.go (hunk 22 above)
Comment (hicqu) [inline util/types/json/serdes.go]: I call `strings.sort()` after that.

**Comment by hicqu** [comment]
PTAL again, thanks.

**Comment by tiancaiamao** [comment]
CI fail. @hicqu

**Comment by hicqu** [comment]
fixed. PTAL again, thanks.

File: util/types/json/compare.go (hunk 13 above)
Comment (tiancaiamao) [inline util/types/json/compare.go]: Really confusing me...
compare(null, false) = ?
compare(null, true) = ?
compare(true, false) = ?

**Comment by shenli** [comment]
@tiancaiamao PTAL

File: util/types/json/compare.go [hunk 23]
```diff
@@ -88,32 +75,33 @@ func CompareJSON(j1 JSON, j2 JSON) (cmp int, err error) {
...
-				cmp, _ = CompareJSON(elem1, elem2)
-				if cmp != 0 {
+		case typeCodeArray:
+			left := j1.array
+			right := j2.array
+			for i := 0; i < len(left) && i < len(right); i++ {
+				elem1 := left[i]
+				elem2 := right[i]
+				cmp, err = CompareJSON(elem1, elem2)
+				if cmp != 0 || err != nil {
```
Comment (shenli) [inline util/types/json/compare.go]: return cmp, errors.Trace(err)

File: util/types/json/compare.go (hunk 13 above)
Comment (shenli) [inline util/types/json/compare.go]: I also wonder compare(null, false) = ?
In most cases, compare a null value with anything else will return null.

File: util/types/json/serdes.go [hunk 24]
```diff
@@ -406,23 +360,34 @@ func (m jsonObject) getSortedKeys() []string {
...
 	} else {
 		var valueOffset = uint32(prefixLen + values.Len())
 		binary.Write(valueEntrys, binary.LittleEndian, valueOffset)
-		value.encode(values)
+		encode(value, values)
 	}
 	return
 }
+
+func pushInlineValue(buffer *bytes.Buffer, value JSON) {
```
Comment (shenli) [inline util/types/json/serdes.go]: Add comment for this function and the padding logic.

File: util/types/json/compare.go (hunk 13 above)
Comment (hicqu) [inline util/types/json/compare.go]: `select null < json` will return null, it's processed in `expression/builtin_compare.go`.
`select json_null < true` will return true.

**Comment by tiancaiamao** [comment]
LGTM
You can add more comment to the code.

File: util/types/json/compare.go (hunk 23 above)
Comment (hicqu) [inline util/types/json/compare.go]: fixed.

**Comment by XuHuaiyu** [comment]
LGTM
